File: app/src/at-trivo/java/asiantech/internship/summer/recycler_view/TimeLineFragment.java
package asiantech.internship.summer.recycler_view;

import android.content.Context;
import android.os.Bundle;
import android.os.Handler;
import android.support.annotation.NonNull;
import android.support.v4.app.Fragment;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.RecyclerView.OnScrollListener;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import java.util.ArrayList;
import java.util.List;

import asiantech.internship.summer.R;
import asiantech.internship.summer.recycler_view.model.Timeline;
import asiantech.internship.summer.recycler_view.model.TimelineCreator;
import asiantech.internship.summer.recycler_view.timeline_recycler_view.TimelineAdapter;
import asiantech.internship.summer.recycler_view.timeline_recycler_view.TimelineViewHolder;

public class TimeLineFragment extends Fragment
        implements TimelineViewHolder.TimeLineViewHolderListener {

    private static final int TIME_DELAY = 5000;
    protected List<Timeline> mDataSet;
    private LinearLayoutManager mLayoutManager;
    protected TimelineAdapter mAdapter;
    private RecyclerView mRecyclerView;
    protected SwipeRefreshLayout mSwipeRefreshLayout;
    private boolean mIsLoading;
    private TimelineFragmentListener mListener;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mDataSet = new ArrayList<>();
        setUpDataSet();
    }

    protected void setUpDataSet() {
        mDataSet.add(null);
        mDataSet.addAll(TimelineCreator.createListTimeline());
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_time_line, container, false);
        mRecyclerView = view.findViewById(R.id.recyclerView);
        mSwipeRefreshLayout = view.findViewById(R.id.swipeRefresh);
        setUpRecyclerView();
        setUpRecyclerViewOnScroll();
        setUpSwipeRefresh();
        return view;
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        if (context instanceof TimelineFragmentListener) {
            mListener = (TimelineFragmentListener) context;
        }
    }

    @Override
    public void onDetach() {
        super.onDetach();
        mListener = null;
    }

    private void setUpRecyclerView() {
        mLayoutManager = new LinearLayoutManager(getActivity());
        mAdapter = new TimelineAdapter(mDataSet, this.getActivity(), this);
        mRecyclerView.setLayoutManager(mLayoutManager);
        mRecyclerView.setAdapter(mAdapter);
    }

    protected void setUpRecyclerViewOnScroll() {
        mRecyclerView.addOnScrollListener(new OnScrollListener() {
            @Override
            public void onScrolled(RecyclerView recyclerView, int dx, int dy) {
                super.onScrolled(recyclerView, dx, dy);
                if (!mIsLoading && mLayoutManager.getItemCount() == mLayoutManager.findLastVisibleItemPosition() + 1) {
                    mIsLoading = true;
                    mDataSet.add(null);
                    mAdapter.notifyItemInserted(mDataSet.size() - 1);
                    new Handler().postDelayed(() -> {
                        mDataSet.remove(mDataSet.size() - 1);
                        mAdapter.notifyItemRemoved(mDataSet.size());
                        mDataSet.addAll(TimelineCreator.createListTimeline());
                        mAdapter.notifyDataSetChanged();
                        mIsLoading = false;
                    }, TIME_DELAY);
                }
            }
        });
    }

    protected void setUpSwipeRefresh() {
        mSwipeRefreshLayout.setOnRefreshListener(() -> {
            mSwipeRefreshLayout.setRefreshing(true);
            new Handler().postDelayed(() -> {
                mDataSet.clear();
                mDataSet.addAll(TimelineCreator.createListTimeline());
                mAdapter.notifyDataSetChanged();
                mSwipeRefreshLayout.setRefreshing(false);
            }, TIME_DELAY);
        });
    }

    @Override
    public void onHeartImageClick(int position) {
        Timeline timeline = mDataSet.get(position);
        timeline.changeLoveState();
        mAdapter.notifyItemChanged(position);
        if (mListener != null) {
            mListener.onHeartImageClick(this.getClass(),position);
        }
    }

    public interface TimelineFragmentListener {
        void onHeartImageClick(Class fragmentClass,int position);
    }
}

File: app/src/at-trivo/java/asiantech/internship/summer/recycler_view_kotlin/TimelineFragment.kt
package asiantech.internship.summer.recycler_view_kotlin

import android.os.Bundle
import android.os.Handler
import android.support.v4.app.Fragment
import android.support.v4.widget.SwipeRefreshLayout
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import asiantech.internship.summer.R
import asiantech.internship.summer.recycler_view_kotlin.model.Timeline
import asiantech.internship.summer.recycler_view_kotlin.model.TimelineCreator
import asiantech.internship.summer.recycler_view_kotlin.timeline_recycler_view.TimelineAdapter

const val TIME_DELAY: Long = 5000

class TimelineFragment : Fragment(), TimelineFragmentListener {
    private val dataset: MutableList<Timeline?> = mutableListOf()

    init {
        dataset.addAll(TimelineCreator.createTimelines())
    }

    private lateinit var recyclerView: RecyclerView
    private lateinit var swipeRefreshLayout: SwipeRefreshLayout
    private val viewLayoutManager = LinearLayoutManager(activity)
    private val viewAdapter = TimelineAdapter(dataset, this)
    private var isLoading: Boolean = false

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.fragment_time_line, container, false)
        recyclerView = view.findViewById(R.id.recyclerView)
        swipeRefreshLayout = view.findViewById(R.id.swipeRefresh)
        setUpRecyclerView()
        return view
    }

    private fun setUpRecyclerView() {
        recyclerView.apply {
            layoutManager = viewLayoutManager
            adapter = viewAdapter
            setHasFixedSize(true)
        }

        recyclerView.addOnScrollListener(object : RecyclerView.OnScrollListener() {
            override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                super.onScrolled(recyclerView, dx, dy)
                if (!isLoading && viewLayoutManager.itemCount == viewLayoutManager.findLastVisibleItemPosition() + 1) {
                    isLoading = true
                    dataset.add(null)
                    viewAdapter.notifyItemInserted(dataset.size - 1)
                    Handler().postDelayed({
                        dataset.removeAt(dataset.size - 1)
                        viewAdapter.notifyItemRemoved(dataset.size)
                        dataset.addAll(TimelineCreator.createTimelines())
                        viewAdapter.notifyDataSetChanged()
                        isLoading = false
                    }, TIME_DELAY)
                }
            }
        })

        swipeRefreshLayout.setOnRefreshListener {
            swipeRefreshLayout.isRefreshing = true
            Handler().postDelayed({
                dataset.clear()
                dataset.addAll(TimelineCreator.createTimelines())
                viewAdapter.notifyDataSetChanged()
                swipeRefreshLayout.isRefreshing = false
            }, TIME_DELAY)
        }
    }

    override fun onHeartImageClicked(position: Int) {
        dataset[position]?.changeState()
        viewAdapter.notifyItemChanged(position)
    }
}
